lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
   return result;
 }
 
+Optional<std::pair<unsigned, unsigned>>
+FIRParser::getFieldIDFromTokens(SMLoc loc, ArrayAttr tokens, Type type) {
+  if (!type)
+    return None;
+  if (tokens.empty())
+    return std::pair<unsigned, unsigned>(0, 0);
+
+  auto currentType = type.cast<FIRRTLType>();
+  unsigned id = 0;
+  unsigned idRange = 0;
+
+  // Construct a string for error emission.
+  SmallString<16> tokensStr;
+  for (auto token : tokens) {
+    if (auto intAttr = token.dyn_cast<IntegerAttr>())
+      intAttr.getValue().toStringSigned(tokensStr);
+    else if (auto strAttr = token.dyn_cast<StringAttr>())
+      tokensStr.append(strAttr.getValue());
+    tokensStr.push_back('_');
+  }
+  tokensStr.pop_back();
+
+  unsigned tokenIdx = 1;
+  for (auto token : tokens) {
+    // Unflip if the current type is flipped type.
+    if (auto flipType = currentType.dyn_cast<FlipType>())
+      currentType = flipType.getElementType();
+
+    if (auto bundleType = currentType.dyn_cast<BundleType>()) {
+      auto subFieldAttr = token.dyn_cast<StringAttr>();
+      if (!subFieldAttr) {
+        emitError(loc, "expect a string for the " + std::to_string(tokenIdx) +
+                           "-th token of " + tokensStr);
+        return None;
+      }
+
+      // Token should point to valid sub-field.
+      auto subField = subFieldAttr.getValue();
+      auto index = bundleType.getElementIndex(subField);
+      if (!index) {
+        emitError(loc, "the " + std::to_string(tokenIdx) + "-th token of " +
+                           tokensStr + " is not found in the bundle");
+        return None;
+      }
+
+      id += bundleType.getFieldID(index.getValue());
+      currentType = bundleType.getElementType(subField);
+      idRange = currentType.getMaxFieldID();
+
+    } else if (auto vectorType = currentType.dyn_cast<FVectorType>()) {
```

Comment:
You can remove the long if/else if/else chain if you move initialize `tokenIdx` to zero, increment it at the start of the loop, and then exit via `continue` in each of the cases. This would produce something possibly more readable of:
```c++
if (foo) {
  // ... 
  continue;
}
if (bar) {
  // ...
  continue;
}
```

Alternatively, this could make use of a three-case `TypeSwitch`.